### PR TITLE
Fix Issue 18755 - std.typecons.Rebindable breaks @safe-ty

### PR DESCRIPTION
**Pull request by edi33416**

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -2086,9 +2086,16 @@ if (is(T == class) || is(T == interface) || isAssociativeArray!T)
...
     {
-        stripped = cast(U) another;
+        // If `T` defines `opCast` we must infer the safety
+        static if (hasMember!(T, "opCast"))
+        {
+            // This will allow the compiler to infer the safety of `T.opCast!U`
+            // without generating any runtime cost
+            if (false) { stripped = cast(U) another; }
+        }
+        () @trusted { stripped = cast(U) another; }();
```
Comment (n8sh) [inline std/typecons.d]: What is the point of keeping this lambda? Is there a reason not to just remove `@trusted` from the function?

Comment (ZombineDev) [reply std/typecons.d]: For the case where there's not `opCast` member. In which case casting would be un-`@safe`, but we trust that it's done correctly.

File: std/typecons.d [hunk 2]
```diff
@@ -2260,6 +2268,23 @@ if (is(T == class) || is(T == interface) || isDynamicArray!T || isAssociativeArr
...
+        auto opCast(T)() @system immutable pure nothrow
+        {
+            *(cast(uint*) 0xdeadbeef) = 0xcafebabe;
+            return T.init;
+        }
+    }
+
+    static assert(!__traits(compiles, () @safe {
+        auto r = Rebindable!(immutable Foo)(new Foo);
+    }));
```
Comment (ZombineDev) [inline std/typecons.d]: Can you also add for completeness:
```D
    static assert(!__traits(compiles, () @system {
        auto r = Rebindable!(immutable Foo)(new Foo);
    }));
```

Comment (edi33416) [reply std/typecons.d]: Done

File: std/typecons.d (hunk 1 above)
Comment (edi33416) [reply std/typecons.d]: I feel that the same should happen for the `opAssign` overloads, that are also `@trusted`, as they are subject to the same issue

If you agree, I'll open a separate PR for those.

**Comment by ZombineDev** [comment]
@n8sh anything to add from your side? I think this PR is good to go.

File: std/typecons.d (hunk 1 above)
Comment (ZombineDev) [reply std/typecons.d]: Yes, we can explore that in a separate PR, but I'm not sure if we can do anything for e.g. this one:

https://github.com/dlang/phobos/blob/dbca8521b5164c2ba95401d9bc784ec653e1a157/std/typecons.d#L2101-L2104

**Comment by n8sh** [comment]
@ZombineDev seems fine to me.

**Pull request merged**